// @typegoose/typegoose/lib/internal/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfContainingClassError = exports.CannotBeSymbol = exports.InvalidWhatIsItError = exports.AssertionFallbackError = exports.NoValidClassError = exports.NotAllVPOPElementsError = exports.NotStringTypeError = exports.NotNumberTypeError = exports.InvalidTypeError = void 0;
const utils_1 = require("./utils");
class InvalidTypeError extends Error {
    constructor(targetName, key, Type) {
        super(`"${targetName}.${key}"'s Type is invalid! Type is: "${Type}" [E009]`);
    }
}
exports.InvalidTypeError = InvalidTypeError;
class NotNumberTypeError extends Error {
    constructor(targetName, key, enumKey, enumValue) {
        super(`Typeof "${targetName}.${key}" is "Number", value is undefined/null or does not have a reverse mapping! [E011]\n` +
            `  Encountered with property: ${enumKey}.${typeof enumValue}`);
    }
}
exports.NotNumberTypeError = NotNumberTypeError;
class NotStringTypeError extends Error {
    constructor(targetName, key, enumKey, enumValue) {
        super(`Typeof "${targetName}.${key}" is "String", used enum is not only Strings! [E010]\n` +
            `  Encountered with property in Enum: ${enumKey}.${typeof enumValue}`);
    }
}
exports.NotStringTypeError = NotStringTypeError;
/** Not All Virtual Populate Elements Error */
class NotAllVPOPElementsError extends Error {
    constructor(name, key) {
        super(`"${name}.${key}" has not all needed Virtual Populate Options! Needed are: ${utils_1.allVirtualoptions.join(', ')} [E006]`);
    }
}
exports.NotAllVPOPElementsError = NotAllVPOPElementsError;
class NoValidClassError extends TypeError {
    constructor(cl) {
        super(`"${cl}" is not a function(/constructor)!`);
    }
}
exports.NoValidClassError = NoValidClassError;
class AssertionFallbackError extends Error {
    constructor() {
        super('Assert failed - no custom error [E019]');
    }
}
exports.AssertionFallbackError = AssertionFallbackError;
/** Error for when an unknown WhatIsIt is passed to an switch, gets thrown in the default case */
class InvalidWhatIsItError extends Error {
    constructor(whatisit, name, key, where) {
        super(`"${whatisit}"(${where}) is invalid for "${name}.${key}" [E013]`);
    }
}
exports.InvalidWhatIsItError = InvalidWhatIsItError;
class CannotBeSymbol extends Error {
    constructor(name, key) {
        super(`A property key in Typegoose cannot be an symbol! (${name}.${String(key)}) [E024]`);
    }
}
exports.CannotBeSymbol = CannotBeSymbol;
class SelfContainingClassError extends TypeError {
    constructor(name, key) {
        super('It seems like the type used is the same as the target class, which is not supported\n' +
            `Please look at https://github.com/typegoose/typegoose/issues/42 for more information (${name}.${key}) [E004]`);
    }
}
exports.SelfContainingClassError = SelfContainingClassError;
